core/src/main/java/com/linecorp/armeria/server/HttpServerCodec.java

```diff
@@ -0,0 +1,151 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.server;
+
+import static io.netty.handler.codec.http.HttpObjectDecoder.DEFAULT_MAX_CHUNK_SIZE;
+import static io.netty.handler.codec.http.HttpObjectDecoder.DEFAULT_MAX_HEADER_SIZE;
+import static io.netty.handler.codec.http.HttpObjectDecoder.DEFAULT_MAX_INITIAL_LINE_LENGTH;
+
+import java.util.ArrayDeque;
+import java.util.List;
+import java.util.Queue;
+
+import com.linecorp.armeria.common.annotation.Nullable;
+import com.linecorp.armeria.server.ServerHttp1ObjectEncoder.WebSocketUpgradeContext;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.CombinedChannelDuplexHandler;
+import io.netty.handler.codec.http.DefaultHttpContent;
+import io.netty.handler.codec.http.HttpHeaderNames;
+import io.netty.handler.codec.http.HttpMethod;
+import io.netty.handler.codec.http.HttpRequest;
+import io.netty.handler.codec.http.HttpRequestDecoder;
+import io.netty.handler.codec.http.HttpResponse;
+import io.netty.handler.codec.http.HttpResponseEncoder;
+import io.netty.handler.codec.http.HttpServerUpgradeHandler;
+import io.netty.handler.codec.http.HttpStatusClass;
+
+final class HttpServerCodec extends CombinedChannelDuplexHandler<HttpRequestDecoder, HttpResponseEncoder>
+        implements HttpServerUpgradeHandler.SourceCodec {
+
+    // Forked from Netty 4.1.69 at 34a31522f0145e2d434aaea2ef8ac5ed8d1a91a0
+    // - Added a logic that does not decode HTTP palyoad when WebSocketSession is established.
+
+    /** A queue that is used for correlating a request and a response. */
+    private final Queue<HttpMethod> queue = new ArrayDeque<>();
+
+    HttpServerCodec() {
+        this(DEFAULT_MAX_INITIAL_LINE_LENGTH, DEFAULT_MAX_HEADER_SIZE, DEFAULT_MAX_CHUNK_SIZE,
+             WebSocketUpgradeContext.noop());
+    }
+
+    /**
+     * Creates a new instance with the specified decoder options.
+     */
+    HttpServerCodec(int maxInitialLineLength, int maxHeaderSize, int maxChunkSize,
+                    WebSocketUpgradeContext webSocketUpgradeContext) {
+        init(new HttpServerRequestDecoder(maxInitialLineLength, maxHeaderSize,
+                                          maxChunkSize, webSocketUpgradeContext),
+             new HttpServerResponseEncoder());
+    }
+
+    /**
+     * Upgrades to another protocol from HTTP. Removes the {@link HttpRequestDecoder} and
+     * {@link HttpResponseEncoder} from the pipeline.
+     */
+    @Override
+    public void upgradeFrom(ChannelHandlerContext ctx) {
+        ctx.pipeline().remove(this);
+    }
+
+    private final class HttpServerRequestDecoder extends HttpRequestDecoder {
```

Comment:
Side note: If we override `HttpRequestDecoder.createMessage()` and provide our own implementation of `HttpMessage`, we can covert Netty HTTP 1 headers to Armeria headers with zero-copy.

Reply:
Nice! It could improve our HTTP/1 services performance.

Reply:
I removed the forked `HttpServerCodec` so I created an issue. 😉 
https://github.com/line/armeria/issues/4853